src/clone_area_cmd.cpp

```diff
@@ -0,0 +1,533 @@
+		dest_point = Rotate(dest_point, angle);
+		dest_point.x += dest_position_x;
+		dest_point.y += dest_position_y;
+
+		TileIndex dest_tile = TileXY(dest_point.x, dest_point.y);
+
+		uint origin_height = TileHeight(iter_tile) + difference_height;
+		uint dest_height = TileHeight(dest_tile);
+		
+		while (dest_height != origin_height) {
+			CommandCost ret;
+			std::tie(ret, std::ignore, error_tile) = Command<CMD_TERRAFORM_LAND>::Do(flags & ~DC_EXEC, dest_tile, SLOPE_N, dest_height <= origin_height);
+			if (ret.Failed()) {
+				last_error = ret;
+
+				if (!terraform_problem) {
+					terraform_problem_tile = error_tile;
+					terraform_error = last_error;
+					terraform_problem = true;
+				}
+
+				/* Did we reach the limit? */
+				if (ret.GetErrorMessage() == STR_ERROR_TERRAFORM_LIMIT_REACHED) limit = 0;
+				break;
+			}
+
+			if (flags & DC_EXEC) {
+				money -= ret.GetCost();
+				if (money < 0) {
+					return { cost, ret.GetCost(), error_tile };
+				}
+				Command<CMD_TERRAFORM_LAND>::Do(flags, dest_tile, SLOPE_N, dest_height <= origin_height);
+			} else {
+				/* When we're at the terraform limit we better bail (unneeded) testing as well.
+				 * This will probably cause the terraforming cost to be underestimated, but only
+				 * when it's near the terraforming limit. Even then, the estimation is
+				 * completely off due to it basically counting terraforming double, so it being
+				 * cut off earlier might even give a better estimate in some cases. */
+				if (--limit <= 0) {
+					had_success = true;
+					break;
+				}
+			}
+
+			cost.AddCost(ret);
+			dest_height += (dest_height > origin_height) ? -1 : 1;
+			had_success = true;
+		}
+		if (limit <= 0) break;
+	}
+
+	if (terraform_problem) {
+		return { terraform_error, 0, terraform_problem_tile };
+	}
+
+	CommandCost ret;
+	std::tie(ret, std::ignore, error_tile) = CmdCloneAreaPasteProperty(flags, tile, start_tile, diagonal);
+	if (ret.Failed()) {
+		last_error = ret;
+	} else {
+		had_success = true;
+	}
+	if (flags & DC_EXEC) {
+		money -= ret.GetCost();
+		if (money < 0) {
+			return { cost, ret.GetCost(), error_tile };
+		}
+	}
+
+	CommandCost cc_ret = had_success ? cost : last_error;
+	return { cc_ret, 0, cc_ret.Succeeded() ? tile : error_tile };
+}
+
+
+/**
+ * Insert structures into the selected area on the map.
+ * @param flags operation to perform
+ * @param tile end tile of rail conversion drag
+ * @param area_start start tile of drag
+ * @param diagonal build diagonally or not.
+ * @return the cost of this operation or an error
+ */
+std::tuple<CommandCost, Money, TileIndex> CmdCloneAreaPasteProperty(DoCommandFlag flags, TileIndex tile, TileIndex area_start, [[maybe_unused]] bool diagonal)
+{
+	if (area_start >= Map::Size()) return { CMD_ERROR, 0, INVALID_TILE };
+
+	CommandCost cost(EXPENSES_CONSTRUCTION);
+	CommandCost last_error(INVALID_STRING_ID);
+	bool had_success = false;
+	bool auto_remove_signals = true;
+	bool signals_copy = true;
```

Comment:
Neither of these are ever changed, they are always true. Unfinished work?

Reply:
there is a lot of code from the existing code base
I don't have time to improve every line of code